yarn/src/main/scala/org/apache/spark/deploy/yarn/security/ConfigurableCredentialManager.scala

```diff
@@ -0,0 +1,158 @@
+    "hbase" -> "org.apache.spark.deploy.yarn.security.HBaseCredentialProvider"
+  )
+
+  // AMDelegationTokenRenewer, this will only be create and started in the AM
+  private var _delegationTokenRenewer: AMDelegationTokenRenewer = null
+
+  // ExecutorDelegationTokenUpdater, this will only be created and started in the driver and
+  // executor side.
+  private var _delegationTokenUpdater: ExecutorDelegationTokenUpdater = null
+
+  def initialize(): Unit = {
+    val providers = ServiceLoader.load(classOf[ServiceCredentialProvider],
+      Utils.getContextOrSparkClassLoader).asScala
+
+    // Filter out credentials in which spark.yarn.security.credentials.{service}.enabled is false.
+    providers.filter { p =>
+      sparkConf.getOption(providerEnabledConfig.format(p.serviceName))
+        .orElse {
+          sparkConf.getOption(deprecatedProviderEnabledConfig.format(p.serviceName)).map { c =>
+            logWarning(s"${deprecatedProviderEnabledConfig.format(p.serviceName)} is deprecated, " +
+              s"using ${providerEnabledConfig.format(p.serviceName)} instead")
+            c
+          }
+        }
+        .getOrElse(defaultCredentialProviders.keySet.find(_ == p.serviceName).isDefined.toString)
+        .toBoolean
+    }.foreach {
+      p => credentialProviders(p.serviceName) = p
+    }
+  }
+
+  /**
+   * Get credential provider for the specified service.
+   */
+  def getServiceCredentialProvider(service: String): Option[ServiceCredentialProvider] = {
+    credentialProviders.get(service)
+  }
+
+  /**
+   * Obtain credentials for specified service.
+   * @return time of next renewal if this service credential is renewable.
+   */
+  def obtainCredentialsFromService(
+      service: String,
+      hadoopConf: Configuration,
+      creds: Credentials): Option[Long] = {
+    getServiceCredentialProvider(service).flatMap { provider =>
+      if (provider.isCredentialRequired(hadoopConf)) {
+        provider.obtainCredentials(hadoopConf, creds)
+      } else {
+        None
+      }
+    }
+  }
+
+  /**
+   * Obtain credentials from all the registered providers.
+   * @return Array of time of next renewal if credential can be renewed.
+   *         Otherwise will return None instead.
+   */
+  def obtainCredentials(hadoopConf: Configuration, creds: Credentials): Array[Option[Long]] = {
+    val timeOfNextRenewals = mutable.ArrayBuffer[Option[Long]]()
+    credentialProviders.values.foreach { provider =>
+      if (provider.isCredentialRequired(hadoopConf)) {
+        timeOfNextRenewals += provider.obtainCredentials(hadoopConf, creds)
+      } else {
+        logDebug(s"Service ${provider.serviceName} does not require a token." +
+          s" Check your configuration to see if security is disabled or not.")
+      }
+    }
+
+    timeOfNextRenewals.toArray
+  }
+
+  def delegationTokenRenewer(hadoopConf: Configuration): AMDelegationTokenRenewer = synchronized {
```

Comment:
This method has that weird singleton-like pattern again. Maybe it should just return a new renewer on every call, since it doesn't seem this class actually has any use for it.